File: taboosearch/src/taboosearch/tsp/TspInitializator.scala
package taboosearch.tsp

import scala.collection.mutable.ListBuffer;
import core._

class TspInitializator(var graph : Graph) extends Initializator[TspSolution] {

	def getSolution(startFrom : Int) : List[Int] = {
		val n = graph.weights.length;
		val route = new ListBuffer[Int]();
		var v = startFrom;
		var nearestV = startFrom;
		route.append(v);
		
		for (times <- 1 until n) {
			var min = Double.MaxValue;
			for (w <- 0 until n) {
				if (graph.weights(v)(w) < min && !route.contains(w)) {
				    min = graph.weights(v)(w);
				    nearestV = w;
				}
			}
			route.append(nearestV);
			v = nearestV;
		}
		
		return route.toList;
	}

	def getInitialGeneration() : TspGeneration = {
		val result = new TspGeneration();
		val route = getSolution(0);
		result += new TspSolution(route);
		return result;
	}
	
}

File: src/taboosearch/tsp/TSPInitializator.java
package taboosearch.tsp;

import java.util.LinkedList;
import java.util.List;

import readers.Graph;
import taboosearch.Initializator;

public class TSPInitializator extends Initializator<TSPSolution, TSPSwapMove, TSPGeneration> {
	private double[][] weights;
	private int n = 0;
	TSPEvaluator evaluator;

	public TSPInitializator(Graph graph, TSPEvaluator evaluator) {	
		this.weights = graph.getWeights();
		this.n = graph.getVertexesNumber();
		this.evaluator = evaluator;
	}
	
	private List<Integer> getSolution(int startFrom) {
		LinkedList<Integer> route = new LinkedList<Integer>();
		int v = startFrom,
			nearestV = startFrom;
		route.add(v);
		
		for (int times = 1; times < n; ++times) {
			double min = Double.MAX_VALUE;
			for (int w = 0; w < this.n; ++w) {
				if (this.weights[v][w] < min && !route.contains(w)) {
				    min = this.weights[v][w];
				    nearestV = w;
				}
			}
			route.add(nearestV);
			v = nearestV;
		}

		return route;
	}

	public TSPGeneration getInitialGeneration() {
		TSPSolution solution = new TSPSolution(this.getSolution(0));
		TSPGeneration result = new TSPGeneration();
		solution.setCost(evaluator.evaluate(solution));
		result.add(solution);
		return result;
	}
}
